nengo_loihi/utils/sparse_matrix.py

```diff
@@ -0,0 +1,126 @@
+import warnings
+
+import numpy as np
+
+from nengo_loihi.compat import nengo_transforms, scipy_sparse
+
+
+def is_nengo_sparse(matrix):
+    return nengo_transforms and isinstance(matrix, nengo_transforms.SparseMatrix)
+
+
+def is_scipy_sparse(matrix):
+    return scipy_sparse and isinstance(matrix, scipy_sparse.spmatrix)
+
+
+def is_sparse(matrix):
+    return is_nengo_sparse(matrix) or is_scipy_sparse(matrix)
+
+
+def expand_matrix(matrix, shape):
+    """Ensure matrix is 2-D of the given shape, make it sparse diagonal if not.
+
+    If the matrix is 0-D or 1-D, ensure it's square and make it a diagonal
+    sparse matrix. Otherwise, just check the shape.
+    """
+    if matrix.ndim < 2:
+        assert shape[0] == shape[1]
+        data = matrix * np.ones(shape[0]) if matrix.ndim == 0 else matrix
+        assert data.size == shape[0]
+        if scipy_sparse:
+            matrix = scipy_sparse.dia_matrix((data, 0), shape=(shape[0], shape[0]))
+        elif nengo_transforms:
+            indices = np.column_stack([np.arange(shape[0]), np.arange(shape[0])])
+            matrix = nengo_transforms.SparseMatrix(
+                indices=indices, data=data, shape=(shape[0], shape[0])
+            )
+        else:
+            if shape[0] ** 2 > 4:
+                warnings.warn(
+                    "Sparse matrices not supported by Nengo < 3.0. Using "
+                    "dense matrices instead. Please upgrade for efficient "
+                    "connections."
+                )
+            matrix = np.diag(data)
+    else:
+        assert matrix.ndim == 2
+
+    assert matrix.shape == shape
+    return matrix
+
+
+def scale_matrix(matrix, scale):
+    scale = np.asarray(scale)
+
+    if is_nengo_sparse(matrix):
+        assert scale.ndim < 2
+
+        data = matrix.data.copy()
+        if scale.size > 1:
+            assert scale.size == matrix.shape[1]
+            for k, s in enumerate(scale):
+                data[matrix.indices[:, 1] == k] *= s
+        else:
+            data *= scale
+
+        return nengo_transforms.SparseMatrix(
+            indices=matrix.indices, data=data, shape=matrix.shape
+        )
+
+    elif is_scipy_sparse(matrix):
+        assert scale.ndim < 2
+
+        if scale.size > 1:
+            assert scale.size == matrix.shape[1]
+            diag = scipy_sparse.dia_matrix((scale, 0), shape=(scale.size, scale.size))
+            return matrix.dot(diag)
+        else:
+            return matrix * scale
+    else:
+        return matrix * scale
+
+
+def stack_matrices(matrices, order="v"):
```

Comment:
An option would be to make `scipy` a dependency of nengo-loihi, then convert all sparse matrices to scipy matrices. That would let us simplify this whole file a lot.  Since nxsdk already requires scipy, this would only make a difference for emulator-only users (and it's not the end of the world for them).

Reply:
Yeah, I'd be fine with that. `nxsdk` doesn't *really* require Scipy I don't think; I'm pretty sure it's just for one of their extra packages. But they way they do their `setup.py` they just put everything they might ever use as "requirements". But that's how everyone is going to be installing. Scipy is also way easier to install than it used to be (on most hardware at least).